client/write.go

```diff
@@ -117,25 +121,54 @@ func (w *writeState) reset() {
 func (w *writeState) close() {
 	w.session.writeOpPool.Put(w.op)
 
-	w.op, w.majority, w.successful, w.pending = nil, 0, 0, 0
+	w.op, w.majority, w.pending, w.halfSuccess = nil, 0, 0, 0
 	w.errors = w.errors[:0]
 	w.queues = w.queues[:0]
 
 	w.session.writeStatePool.Put(w)
 }
 
+func (w *writeState) addError(err error) {
+	w.Lock()
+	w.errors = append(w.errors, err)
+	w.Unlock()
+}
+
+// todo@bl - modify comment on what happens to inflight writes
 func (w *writeState) completionFn(result interface{}, err error) {
 	var (
 		successful int32
 		remaining  = atomic.AddInt32(&w.pending, -1)
+		hostID     = result.(string)
```

Comment:
The host queue doesn't currently know anything about shards, which I think is the correct separation of concerns. If we wanted to get the shard state or a score earlier, the host queue would need to inspect the writeOp for the shardID and contain logic on getting and scoring shard states. 

Alternatively, we could also add a "score" function to the writeOp that takes a hostID and use that from the host queue, but that's needlessly roundabout.